<a href="https://colab.research.google.com/github/bunkorner/VR_Assignment2/blob/main/Assignment_2b.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import cv2
import os
import numpy as np

In [2]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [3]:
srcPath = '/content/gdrive/MyDrive/CV/Assignment_2/Assignment2_BikeHorses'
destPath = '/content/gdrive/MyDrive/CV/Assignment_2/Bike_Horse'

Organized the Dataset, which previously had IMG1_(Copy).jpg is now IMG1.jpg, IMG2.jpg and so on.

Taking all the images from Source Path and making an organized copy in Destination Path

In [4]:
if not os.path.exists(destPath):
    os.mkdir(destPath)

In [5]:
onlyDirs = [f for f in os.listdir(srcPath) if os.path.isdir(os.path.join(srcPath, f))]

for directory in onlyDirs:
    rdPath = srcPath+'/'+directory
    path = destPath+'/'+directory
    if not os.path.exists(path):
        os.makedirs(path)
        
    onlyfiles = [f for f in os.listdir(rdPath) if os.path.isfile(os.path.join(rdPath, f))]
    i=0
    for file in onlyfiles:
        i = i + 1
        imgPath = rdPath+'/'+file
        img = cv2.imread(imgPath)
        wrtPath = path+'/' + str(i) + '.jpg'
        cv2.imwrite(wrtPath,img)

Imgs will store the Path along with Name of all the images.

Labels will store the Ground Truth for the respective image.

In [6]:
imgs = []
labels = []
label = -1

dirs = [f for f in os.listdir(destPath) if os.path.isdir(os.path.join(destPath, f))]

for directory in dirs:
    label = label + 1
    path = destPath+'/'+directory
    files = [f for f in os.listdir(path) if os.path.isfile(os.path.join(path, f))]

    for file in files:
        imgPath = path+'/'+file
        imgs.append(imgPath)
        labels.append(label)

In [7]:
len(imgs)

188

There are 188 images in total.

In [8]:
imgFiles = []
for imgName in imgs:
    img = cv2.imread(imgName)
    imgFiles.append(img)
imgFiles = np.asarray(imgFiles)

<ipython-input-8-eb5b51ae2ef9>:5: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  imgFiles = np.asarray(imgFiles)


ImgFiles will contain all the data of images stacked one over other.

In [9]:
from sklearn.cluster import KMeans
numberOfClusters=300

We have kept K = 300 here, that is there will be 300 unique visual words.

In [10]:
descriptors = np.asarray([])
ext = cv2.SIFT_create()
for img in imgFiles:
    gray = cv2.cvtColor(img,cv2.COLOR_RGB2GRAY)
    keypoints, desc = ext.detectAndCompute(gray, None)
    if type(desc) == np.ndarray :
        if descriptors.shape[0] == 0:
            descriptors = desc
        else:
            descriptors = np.concatenate((descriptors, desc), axis=0)
kmeans = KMeans(n_clusters = numberOfClusters)

Using SIFT to extract features from all the image.

Now, those extracted features in terms of patches is fed to KMeans algorithm.

In [11]:
descriptors.shape

(178044, 128)

Here, we can see there are 178044 different features obtained from 188 images, these features will have many repeating (similar) features such as wheels, Face of the horse and so on, which will be then considered as a single word.

In [12]:
kmeans.fit(descriptors)

KMeans(n_clusters=300)

Now all the features are divided into 300 clusters.

In [13]:
allBOVW = []
ext = cv2.xfeatures2d.SIFT_create()
for img in imgFiles:
  bovw = [0]*numberOfClusters
  gray = cv2.cvtColor(img,cv2.COLOR_RGB2GRAY)
  keypoints, descriptors = ext.detectAndCompute(gray, None)
  if type(descriptors) == np.ndarray:
      pred = kmeans.predict(descriptors)
      for ele in pred:
          bovw[ele] = bovw[ele] + 1
  allBOVW.append(bovw)

When we have a new image, we again extract its features and descriptors using SIFT, then feed this as input to our now trained KMeans model, then with the predicted word we'll create a histogram (named bovw, here).

This histogram will now be useful as Train_X and Labels as our Train_Y.

In this piece of code, we find the histogram for all the images (i.e. entire ImgFiles array)

In [14]:
allBOVW = np.asarray(allBOVW)

In [15]:
from sklearn.model_selection import train_test_split

Now we split the data into Training and Testing data, to know how our models performed

In [16]:
trainImgs,testImgs, trainLabels, testLabels = train_test_split(allBOVW, labels, train_size=0.7, random_state=0)

In [17]:
trainImgs

array([[ 1,  0,  1, ...,  1,  2,  1],
       [ 0,  0,  0, ...,  0,  0,  0],
       [ 5,  0, 16, ...,  5,  5,  1],
       ...,
       [ 1,  0,  6, ...,  0,  2,  3],
       [ 2,  0,  2, ...,  3,  4,  6],
       [ 0,  0,  3, ...,  3,  3,  2]])

Using Logistic Regression

In [22]:
from sklearn.linear_model import LogisticRegression
LRegression = LogisticRegression().fit(trainImgs, trainLabels)

In [23]:
LRegression.score(testImgs,testLabels)

0.9649122807017544

Using Support Vector Classifier

In [20]:
from sklearn import svm
SVClassifier = svm.SVC(gamma=0.01,kernel='linear')
SVClassifier.fit(trainImgs,trainLabels)

SVC(gamma=0.01, kernel='linear')

In [21]:
SVClassifier.score(testImgs,testLabels)

0.9649122807017544

Tried running the same Code but instead of using SIFT, I tried SURF and ORB, the code took less time to execute but there was no loss in Accuracy as such.